docs/testing-strategy.md

```diff
@@ -0,0 +1,154 @@
+breaking behavioural changes; and for it to be easy for developers to know
+where to add tests. The purpose of this document is
+to provide a summary of the testing approaches used so that new and old
+developers to Content Publisher have a shared understanding.
+
+The principles we most value in testing are:
+
+- consistency across tests over developer preference - all tests should
+  feel like they could have been written by the same person;
+- readable tests that are easy to follow - aim for expectations to be clear,
+  setup code to be close to associated assertions and avoid being overzealous
+  in removing repetition;
+- avoid testing configuration - tests should be focused on testing logic we've
+  introduced, rather than whether or not our code was typed correctly;
+- pragmatism in what we test - if all code paths of a class/module are
+  exhausted by other tests then we may not need to explicitly test something
+  further.
+
+This testing strategy was inspired by:
+[Whitehall Testing Guidelines][whitehall-tests],
+[Email Alert API Testing][email-alert-api-tests], and
+[Thoughtbots' How We Test Rails Applications][thoughtbot-rails-tests]
+
+## Tests directory structure
+
+Tests for Content Publisher are stored in the [spec](../spec) directory with
+the Ruby tests written using [RSpec](https://rspec.info/) and JavaScript tests
+written using [Jasmine](https://jasmine.github.io/).
+
+As per common Rails conventions, most of the directories within the spec
+directory contain Ruby [unit tests](#unit-tests) with the directories matching
+those within the app directory. Code that is stored in the lib directory has
+corresponding tests in the spec/lib directory.
+
+Directories within spec that don't contain Ruby unit tests are as follows:
+
+- factories - contains [FactoryBot factories][factory-bot], which are a fixture
+  alternative;
+- fixtures - contains supporting files for tests (for example image files) which
+  aren't easily produced by factories;
+- features - contains [feature tests](#feature-tests), which exercises
+  application features from a user perspective;
+- javascripts - contains [unit tests](#unit-tests) for JavaScript files
+  and configuration and helper files for Jasmine;
+- requests - contains [request tests](#request-tests), which test the HTTP API
+  of the application;
+- support - contains helper files for shared test methods and logic;
+- views - contains [view tests](#view-tests), which are used to test
+  [Rails views][rails-views] output expected HTML in a given scenario.
+
+## Unit tests
+
+The purpose of these tests is to test individual units of the system in
+isolation. They are intended to provide exhaustive tests of the code paths
+through a particular class or module through its public interface. As per
+the [test pyramid](https://martinfowler.com/bliki/TestPyramid.html) approach
+to software testing, these should provide the majority of test coverage.
```

Comment:
Maybe we're on a slightly different wavelength here. I can see this is somewhat ambiguous based on how the word coverage is understood. I've amended this to instead reference the volume of tests.

Reply:
Thanks for rephrasing - that is clearer. Perhaps it's just me, but when I read this I get the impression that I "should" write unit tests for everything so that we're "following" the pyramid model. Looking at it again: what point are we trying to communicate in this sentence?

Reply:
Sure - I see what you mean and writing unit tests for "everything" is frequently impractical or lacks pragmatism. More so that unit tests should be the primary mechanism to test logic that we've wrote, and we should be using that to explore the distinct logical paths that code takes thus producing the volume of tests. Since this offers advantages in terms of ease of authoring, ease of debugging and speed of execution compared to higher level tests.

I guess the part of the testing pyramid part description from Martin Fowler that resonates in reference to this is:

"I always argue that high-level tests are there as a second line of test defense. If you get a failure in a high level test, not just do you have a bug in your functional code, you also have a missing or incorrect unit test. Thus I advise that before fixing a bug exposed by a high level test, you should replicate the bug with a unit test. Then the unit test ensures the bug stays dead."

Reply:
Oh dear, I find myself disagreeing with this last paragraph even more: a failure in a high level test doesn't doesn't to me suggest a test is missing, since the existing tests are clearly serving their purpose. The argument is very strange.

I did wonder about simply suggesting "will provide" instead of "should provide", but that's getting too subtle and subjective.

I think we're both in agreement about the pros and cons of (starting with) unit testing. For me, the important question is: what is the *easiest* way to test this? If there's no unit test for something, then writing a full one may be *harder* and possibly more resource intensive than simply modifying a higher-level test.

Let's wait and see if the way I'm thinking about tests rubs against the strategy. Then we can have a more focussed discussion.

Reply:
ha - oh dear indeed. 

I'd suggest that the high level test is serving its purpose in telling you that the functionality is not working however the nature of high level tests shouldn't be telling you which unit of code is not working so you have a gap. If you're lucky enough to have an exception be the problem then you can work out the unit easily enough but otherwise you're digging into UI assertions.

But yes, let us chat more if this becomes actual pain points.